subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractMutableModuleComponentResolveMetadata.java

```diff
@@ -106,7 +108,7 @@ public String getStatus() {
 
     @Override
     public ImmutableMap<String, ? extends ConfigurationMetadata> getConfigurations() {
-        if (configurations == null) {
+        if (configurations == null || mutableDependencies != null) {
```

Comment:
This is due to the fact that as soon as we give out a mutable list, this list might change and `configurations` is potentially always outdated.

As alternative, we can remove the commit https://github.com/gradle/gradle/pull/3158/commits/d2e1a6229600ae78e20aa95b3a4336cc629662cc again and go back to explicitly replacing the whole dependency list instead of lazyly providing a mutable list.